File: 0234-palindrome-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        // vector<int> ans;
        // while(head != NULL){
        //     ans.push_back(head->val);
        //     head = head->next;
        // }
        // int left = 0, right =ans.size() -1;
        // while(left <= right){
        //     if(ans[left++] != ans[right--]){
        //         return false;
        //     }
        // }
        // return true;
        stack<int> ans;
        ListNode* temp = head;
        while(temp != NULL){
            ans.push(temp->val);
            temp = temp->next;
        }
        temp = head;
        while(temp!=NULL){
            if(ans.top() != temp->val)
                return false;
            ans.pop();
            temp = temp->next;
        }
        return true;
    }
};

File: 0234-palindrome-linked-list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode getmiddle(ListNode root){
        ListNode fast = root;
        ListNode slow = root;
        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    } 
    public ListNode reverse(ListNode root){
        ListNode temp = root;
        ListNode prev = null;
        while(temp != null){
            ListNode cur = temp;
            temp = temp.next;
            cur.next = prev;
            prev = cur;
        }
        return prev;
    }
    public void printList(ListNode root){
        ListNode cur = root;
        while(cur != null){
            System.out.print(cur.val + " ");
            cur =cur.next;
        }
        System.out.println();
    }
    public boolean isPalindrome(ListNode head) {
        ListNode middle = getmiddle(head);
        ListNode rev = reverse(middle);
        // printList(rev);
        middle.next = null;
        ListNode temp = rev;
        ListNode temp2 = head;
        while(temp != null && temp2 !=null){
            if(temp.val != temp2.val){
                return false;
            }
            temp  = temp.next;
            temp2 = temp2.next;
        }
        return true;
    }
}
